core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -0,0 +1,717 @@
+
+So we mostly ignore coercions. The one place that I know of where the mismatch hurts
+us is that GHC uses the `Any` type (coercible to and from anything) for unconstrained
+function variables, e.g. in polymorphic lambdas. This is annoying for us, since we
+really want the version with explicit type abstraction. I don't currently have a fix
+for this.
+-}
+
+{- Note [Iterated abstraction and application]
+PLC doesn't expose iterated abstraction and application.
+
+We typically build these with a fold.
+- Iterated application uses a *left* fold, since we want to apply the first variable
+first.
+- Iterated abstraction uses a *right* fold, since we want to bind the first
+variable *last* (so it is on the outside, so will be first when applying).
+-}
+
+-- Useful synonyms and functions
+
+type PCExpr = PC.Term PC.TyName PC.Name ()
+type PCType = PC.Type PC.TyName ()
+
+type PrimTerms = Map.Map GHC.Name (Quote PCExpr)
+type PrimTypes = Map.Map GHC.Name (Quote PCType)
+
+type ConvertingState = (GHC.DynFlags, PrimTerms, PrimTypes, ScopeStack)
+-- See Note [Scopes]
+type Converting m = (Monad m, MonadError (Error ()) m, MonadQuote m, MonadReader ConvertingState m)
+
+strToBs :: String -> BSL.ByteString
+strToBs = BSL.fromStrict . TE.encodeUtf8 . T.pack
+
+bsToStr :: BSL.ByteString -> String
+bsToStr = T.unpack . TE.decodeUtf8 . BSL.toStrict
+
+sdToTxt :: (MonadReader ConvertingState m) => GHC.SDoc -> m T.Text
+sdToTxt sd = do
+  (flags, _, _, _) <- ask
+  pure $ T.pack $ GHC.showSDoc flags sd
+
+conversionFail :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+conversionFail = (throwError . ConversionError) <=< sdToTxt
+
+unsupported :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+unsupported = (throwError . UnsupportedError) <=< sdToTxt
+
+freeVariable :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+freeVariable = (throwError . FreeVariableError) <=< sdToTxt
+
+-- Names and scopes
+
+{- Note [Scopes]
+We need a notion of scope, because we have to make sure that if we convert a GHC
+Var into a variable, then we always convert it into the same variable, while also making
+sure that if we encounter multiple things with the same name we produce fresh variables
+appropriately.
+
+So we have the usual mechanism of carrying around a stack of scopes.
+-}
+
+data Scope = Scope (Map.Map GHC.Name (PC.Name ())) (Map.Map GHC.Name (PC.TyName ()))
+type ScopeStack = NE.NonEmpty Scope
+
+initialScopeStack :: ScopeStack
+initialScopeStack = pure $ Scope Map.empty Map.empty
+
+lookupName :: Scope -> GHC.Name -> Maybe (PC.Name ())
+lookupName (Scope ns _) n = Map.lookup n ns
+
+{- Note [PLC names]
+We convert names from Haskell names quite frequently here, but PLC admits a much
+smaller set of valid identifiers. We compromise by mangling the identifier, but
+in the long run it would be nice to have a more principled encoding so we can
+support unicode identifiers as well.
+-}
```

Comment:
Things like this ought to be added as YT tickets, so we make sure not to forget addressing it.

Reply:
I didn't do this because I thought it wasn't actually an issue - we already have a ticket for unicode identifiers. And in fact I can relax this a bit now we're allowed to start with capitals.

Reply:
Ok.